src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+
+		case NBD_RX_NEG_REP_INFO:
+			rx = nbd_rx_neg_rep_info;
+			req_len = nbd->reply_info_length;
+			break;
+
+		case NBD_RX_TRANS_REP_CMD:
+			rx = nbd_rx_trans_rep_cmd;
+			req_len = sizeof ( nbd->rx_trans_reply );
+			break;
+
+		case NBD_RX_TRANS_DATA:
+			rc = nbd_rx_trans_data ( nbd, iobuf );
+			goto done;
+
+		default:
+			assert ( 0 );
+			rc = -EINVAL;
+			goto done;
+		}
+
+		DBGCIO ( nbd, "NBD %p state %d req %zd, off %zd, \n",
+			 nbd, nbd->rx_state, req_len, nbd->rx_offset );
+
+		part_len = req_len - nbd->rx_offset;
+		if ( part_len > iob_len ( iobuf ) )
+			part_len = iob_len ( iobuf );
+		memcpy ( nbd->rx_buffer + nbd->rx_offset,
+			 iobuf->data, part_len );
+		nbd->rx_offset += part_len;
+
+		iob_pull ( iobuf, part_len );
+
+		/* If all the data for this state has not yet been
+		 * received, stay in this state for now.
+		 */
+		if ( nbd->rx_offset != req_len ) {
+			rc = 0;
+			goto done;
+		}
+
+		rc = rx ( nbd );
+		if ( rc ) {
+			DBGC ( nbd, "NBD %p could not process received "
+			       "data: %s\n", nbd, strerror ( rc ) );
+			goto done;
+		}
+
+		nbd->rx_offset = 0;
+	}
+
+ done:
+	/* Free I/O buffer */
+	free_iob ( iobuf );
+
+	/* Destroy session on error */
+	if ( rc != 0 )
+		nbd_close ( nbd, rc );
+
+	return rc;
+}
+
+/**
+ * Handle data transfer window change
+ *
+ * @v nbd		NBD session
+ */
+static void nbd_socket_window_changed ( struct nbd_session *nbd ) {
+	DBGCP ( nbd, "NBD %p socket window changed\n", nbd );
+	nbd_tx_resume ( nbd );
+}
+
+/** NBD socket interface operations */
+static struct interface_operation nbd_socket_op[] = {
+	INTF_OP ( xfer_deliver, struct nbd_session *, nbd_socket_deliver ),
+	INTF_OP ( xfer_window_changed, struct nbd_session *,
+		  nbd_socket_window_changed ),
+	INTF_OP ( intf_close, struct nbd_session *, nbd_close ),
+};
+
+/** NBD socket interface descriptor */
+static struct interface_descriptor nbd_socket_desc =
+	INTF_DESC ( struct nbd_session, socket, nbd_socket_op );
+
+
+/**
+ * Open NBD URI
+ *
+ * @v parent		Parent interface
+ * @v uri		URI
+ * @ret rc		Return status code
+ */
+static int nbd_open ( struct interface *parent, struct uri *uri ) {
+	struct nbd_session *nbd;
+	char *name;
+	struct sockaddr_tcpip server;
+	int rc;
+
+	/* Sanity check */
+	if ( ! uri->host )
+		return -EINVAL;
+	if ( ! uri->path )
+		return -EINVAL;
+
+	/* Allocate and initialise structure */
+	nbd = zalloc ( sizeof ( *nbd ) );
+	if ( ! nbd )
+		return -ENOMEM;
+
+	ref_init ( &nbd->refcnt, nbd_free );
+	intf_init ( &nbd->block, &nbd_block_desc, &nbd->refcnt );
+	intf_init ( &nbd->socket, &nbd_socket_desc, &nbd->refcnt );
+
+	nbd->uri = uri_get ( uri );
+	name = strchr ( uri->path, '/' );
+	if ( name )
+		nbd->export_name = name + 1;
+	else
+		nbd->export_name = "";
+
+	DBGC ( nbd, "NBD %p open %s (%s)\n", nbd, uri->host, nbd->export_name );
+
+	/* Use NBD_OPT_EXPORT_NAME option by default */
+	nbd->use_opt_go = 0;
+
+	if ( strstr ( uri->query, "use-opt-go" ) )
+		nbd->use_opt_go = 1;
```

Comment:
What does `use-opt-go` do, and is it necessary to support this in iPXE?

Reply:
The NBD protocol has two options for starting the transmission phase: older NBD_OPT_EXPORT_NAME and newer NBD_OPT_GO. If a server stops supporting NBD_OPT_EXPORT_NAME, there is an option to use a newer method.

Reply:
Is there an RFC or equivalent document that requires the server to support the older format?

If not, is it viable for iPXE to try the newer format and fall back to trying the older format, so that the user doesn't have to be aware of this detail?  I try to avoid imposing that sort of burden on the user wherever possible

Reply:
As far as I know, there is no such document. The nbd server older than 3.15 doesn't  support newer option. Fallbask is possible, but requires close and reopen of connection... I'd rather improve it in the future.